README
LizardFS is a highly reliable, scalable and efficient distributed file system. It spreads data over a number of physical servers, making it visible to an end user as a single file system.

Useful links:
    - Main code repository - http://cr.skytechnology.pl:8081/lizardfs,
    - GitHub repository mirror - https://github.com/lizardfs/lizardfs,
    - Code review system - http://cr.skytechnology.pl:8081/,
    - Continuous integration system - http://jenkins.lizardfs.org/.

LizardFS is a fork of MooseFS.

More info can be found at www.lizardfs.com.


NEWS
This file lists noteworthy changes in LizardFS.

 * Lizardfs (3.13.0-rc1) (2018-06-28)
 - (all) uRaft HA
 - (all) fixes to EC handling
 - (all) nfs-ganesha plugin changed to use only C code
 - (mount) reduced number of secondary groups retrievals (better performance)
 - (mount) add fuse3 client (better performance, writeback cache)
 - (all) many fixes

 * Lizardfs (3.12.0-rc1) (2017-11-22)
 - (all) C API
 - (all) nfs-ganesha plugin
 - (all) RichACL support (which includes NFSv4)
 - (all) OSX ACL support
 - (master, mount) file lock fixes
 - (mount) client readahead enabled by default
 - (mount) AVX2 extensions support for erasure code goals
 - (chunkserver) more flexible options
 - (all) many fixes

 * Lizardfs (3.11.3) (2017-07-13)
 - (master) fix issues with reporting defective files
 - (mount) fix request size in read cache for empty results

 * Lizardfs (3.11.0) (2017-05-09)
 - (master) improve ACL implementation
 - (master) add option to avoid same-ip chunkserver replication
 - (master) add minimal goal configuration option
 - (master) reimplement directory entry cache for faster lookups
 - (master) add whole-path lookups
 - (master, chunkserver) add chunkserver load awareness
 - (mount) add readahead to improve sequential read perfromance
 - (mount) add secondary groups support
 - (tools) add correct-only flag to filerepair
 - (tools) add -s and -i options to snapshot command
 - (tools) add recursive remove operations (for removing large directories and snapshots)
 - (tools) add tool for stopping execution of tasks (snapshot, recursive remove, etc.)
 - (all) change to semantic versioning system
 - (all) many fixes

 * Lizardfs (3.10.4) (2016-10-07)
 - (master) task manager performance improvements
 - (master) trash fixes

 * Lizardfs (3.10.2) (2016-08-30)
 - (master) redesign in-memory representation of file system objects - at least 30% reduction in RAM usage
 - (master) name storage - a possibility to keep all file names in BerkeleyDB, thus saving even more RAM
 - (master) redesign of trash - increased performance, reduced RAM usage and CPU pressure
 - (master) huge boost of responsiveness - lengthy operations split into asynchronous bits
 - (master) OPERATIONS_DELAY* config entries, which allow postponing metadata operations on restart/disconnect
 - (master) fix improper handling of endangered chunks
 - (chunkserver) memory optimizations - at least 60% reduction in RAM usage
 - (chunkserver) introduce smart descriptor management
 - (tools) brand new `lizardfs` command, a unified replacement for mfs* tools with prompt and bash completion
 - (all) various fixes and community requests

 * Lizardfs (3.10.0-rc) (2016-03-14)
  - (all) Added erasure code goals
  - (all) Added per directory quotas
  - (all) Improved interaction with legacy version (chunkservers, mounts)
  - (all) Ports for OSX and FreeBSD
  - (all) Many fixes

 * Lizardfs (3.9.4) (2015-12-02)
  - (master) Removed master server overload on restarting chunkservers
  - (master) Improved global file locks engine
  - (chunkserver) Fixed leaking descriptors problem
  - (chunkserver) Improved mechanism of moving chunks to new directory layout
  - (chunkserver) Fixed issues related to scanning directories with new chunk format present
  - (mount) Removed hang in mount when chunkserver reported no valid copies of a file
  - (master) Changed handling of legacy (pre-3.9.2) chunkservers in new installations
  - (cgi) Added XOR replication to statistics
  - (all) Removed default linking to tcmalloc library due to performance drop

 * Lizardfs (3.9.2) (2015-10-23)

  - (all) Introduced XOR goal types
  - (all) Added file locks (flock & fcntl)
  - (all) Increased max number of files from 500 million to over 4 billion
  - (all) Introduced managing open file limits by PAM
  - (master) Improved consistency of applying changelogs by shadow masters
  - (master) Redesigned snapshot execution in master
  - (master) Redesigned chunk loop logic
  - (master) Removed hard coded connection limit
  - (master) Added option to limit chunk loop's CPU usage
  - (chunkserver) Added new network threads responsible for handling requests sent by chunkserver's clients
  - (chunkserver) Introduced new more efficient directory layout
  - (chunkserver) Added option to choose if fsync should be performed after each write for increased safety
  - (chunkserver) Removed hard coded connection limit
  - (chunkserver) Added replication network bandwidth limiting
  - (mount) Improved symlink cache and added configurable timeout value
  - (all) Minor bug fixes and improvements

 * Lizardfs (2.6.0) (2015-02-09)

  - (all) Added comments in all config files
  - (all) Improve messages printed by daemons when starting
  - (cgi) A new chunkserver's chart: number of chunk tests
  - (cgi) Fixed paths to static content
  - (cgi) New implementation of the CGI server; mfscgiserv is now deprecated.
  - (cgi) New table: 'Metadata Servers' in the 'Servers' tab
  - (chunkserver) Allowed starts with damaged disks
  - (chunkserver) A new option: HDD_ADVISE_NO_CACHE
  - (chunkserver) Improved handling of disk read errors
  - (chunkserver) Removed 'testing chunk: xxxxxx' log messages
  - (master) A new feature: disabling atime updates (globally)
  - (master) Added possibility to set priority on handling endangered chunks
  - (master) Fixed rotating changelogs and downloading files in shadow mode
  - (master) Fixed handling replication delay on chunkserver disconnection
  - (probe) New commands
  - (probe) Renamed to lizardfs-admin
  - (all) Minor bug fixes and improvements

 * Lizardfs (2.5.4) (2014-11-07)

  - (all) Boost is no longer required to build the source code of LizardFS or use the binary version
  - (all) Added tiering (aka 'custom goal') feature, which allows users to label chunkservers and to request chunks to be stored on specific groups of servers
  - (cgi) "Exports" tabs renamed to "Config", now it also shows goal definitions
  - (cgi) Added new tab "Chunks"

mfscommon/md5.c
/*
   Copyright 2005-2010 Jakub Kruszona-Zawadzki, Gemius SA.

   This file is part of MooseFS.

   MooseFS is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, version 3.

   MooseFS is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with MooseFS.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <string.h>
#include <inttypes.h>
#include "md5.h"

#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

static const uint8_t padding[64] = {
	0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

#define FF(a, b, c, d, x, s, ac) { \
 (a) += F ((b), (c), (d)) + (x) + (uint32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G ((b), (c), (d)) + (x) + (uint32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H ((b), (c), (d)) + (x) + (uint32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I ((b), (c), (d)) + (x) + (uint32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }

static void md5_encode (uint8_t *output,const uint32_t *input,uint32_t len) {
	uint32_t i,j;

	for (i=0,j=0 ; i<len ; i++,j+=4) {
		output[j] = (uint8_t)(input[i]&0xff);
		output[j+1] = (uint8_t)((input[i]>>8)&0xff);
		output[j+2] = (uint8_t)((input[i]>>16)&0xff);
		output[j+3] = (uint8_t)((input[i]>>24)&0xff);
	}
}

static void md5_decode (uint32_t *output,const uint8_t *input,uint32_t len) {
	uint32_t i,j;

	for (i=0,j=0 ; i<len ; i++,j+=4) {
		output[i] = ((uint32_t)input[j]) | (((uint32_t)input[j+1])<<8) | (((uint32_t)input[j+2])<<16) | (((uint32_t)input[j+3])<<24);
	}
}

static void md5_transform (uint32_t state[4],const uint8_t block[64]) {
	uint32_t a=state[0], b=state[1], c=state[2], d=state[3], x[16];

	md5_decode(x,block,16);

	FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
	FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
	FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
	FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
	FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
	FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
	FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
	FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
	FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
	FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
	FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
	FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
	FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
	FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
	FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
	FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

	GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
	GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
	GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
	GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
	GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
	GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
	GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
	GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
	GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
	GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
	GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
	GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
	GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
	GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
	GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
	GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

	HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
	HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
	HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
	HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
	HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
	HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
	HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
	HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
	HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
	HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
	HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
	HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
	HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
	HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
	HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
	HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

	II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
	II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
	II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
	II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
	II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
	II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
	II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
	II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
	II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
	II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
	II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
	II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
	II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
	II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
	II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
	II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

	state[0] += a;
	state[1] += b;
	state[2] += c;
	state[3] += d;

	memset((char*)x,0,sizeof(x));
}



void md5_init (md5ctx *ctx) {
	ctx->count[0] = ctx->count[1] = 0;
	ctx->state[0] = 0x67452301U;
	ctx->state[1] = 0xefcdab89U;
	ctx->state[2] = 0x98badcfeU;
	ctx->state[3] = 0x10325476U;
}

void md5_update (md5ctx *ctx,const uint8_t *buff,uint32_t leng) {
	uint32_t i,indx,partleng;

	indx = (uint32_t)((ctx->count[0]>>3)&0x3F);

	if ((ctx->count[0]+=((uint32_t)leng<<3)) < ((uint32_t)leng<<3)) {
		ctx->count[1]++;
	}
	ctx->count[1] += ((uint32_t)leng>>29);

	partleng = 64-indx;

	if (leng>=partleng) {
		memcpy((char*)(ctx->buffer+indx),(const char*)buff,partleng);
		md5_transform(ctx->state,ctx->buffer);

		for (i=partleng ; i+63<leng ; i+=64) {
			md5_transform(ctx->state,buff+i);
		}
		indx = 0;
	} else {
		i = 0;
	}

	memcpy((char*)(ctx->buffer+indx),(const char*)(buff+i),leng-i);
}

void md5_final (uint8_t digest[16],md5ctx *ctx) {
	uint8_t bits[8];
	uint32_t indx,padleng;

	md5_encode(bits,ctx->count,2);
	indx = (uint32_t)((ctx->count[0]>>3)&0x3f);
	padleng = (indx<56)?(56-indx):(120-indx);
	md5_update (ctx,padding,padleng);
	md5_update(ctx,bits,8);
	md5_encode(digest,ctx->state,4);
	memset((char*)ctx,0,sizeof(md5ctx));
}
